### PR TITLE
Azure OpenAI LLM implementation

### PR DESCRIPTION
**Pull request by MichaelAnckaert**
## Problem

Canopy can't be used with the Azure OpenAI offering.

## Solution

This PR creates a new AzureOpenAILLM class (inherits from OpenAILLM) to be used with the Azure OpenAI offering. 

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Infrastructure change (CI configs, etc)
- [ ] Non-code change (docs, etc)
- [ ] None of the above: (explain here)

## Test Plan

No tests added yet.

### CODE REVIEW

File: src/canopy_server/app.py
Comment (miararoy) [inline src/canopy_server/app.py:340]: The original OpenAI client is automatically gettitng the env, isn't the AzureOpenAIClient doing the same? if so we prefer handling it this way

Comment (miararoy) [reply src/canopy_server/app.py:340]: seems it does:
https://github.com/openai/openai-python/blob/main/src/openai/lib/azure.py#L160

File: src/canopy/llm/openai.py
Comment (miararoy) [inline src/canopy/llm/openai.py:243]: Note that the new openaiclient is automatically getting the AZURE_OPENAI_API_KEY param
https://github.com/openai/openai-python/blob/main/src/openai/lib/azure.py#L160

File: src/canopy_server/app.py
Comment (miararoy) [inline src/canopy_server/app.py:343]: Config loader should automatically handle values, no need to have it here, in fact this will block other LLM implementations

**Review by miararoy — CHANGES_REQUESTED** [review]
Overall direction seems good, main changes is:
- env loading should be handled by client or by explicit passing parameters
- llm object creation in app.py
- adding proper docstrings and adding readme

**Comment by MichaelAnckaert** [comment]
> Overall direction seems good, main changes is:
> 
>     * env loading should be handled by client or by explicit passing parameters
> 
>     * llm object creation in app.py
> 
>     * adding proper docstrings and adding readme

Thanks for your review @miararoy Happy to know the direction is the correct one. 
I'll address your comments and update the PR.

File: src/canopy/llm/openai.py
Comment (MichaelAnckaert) [reply src/canopy/llm/openai.py:244]: From the Azure documentation this should be a date formatted version.
I'll specified a default version here as the OpenAI SDK requires a version to be set, either via parameter or via environment variable. It seemed a bit excessive to me to require the user to specify yet another environment variable. But willing to change if there's another point of view. 

Ref: https://learn.microsoft.com/en-us/azure/ai-services/openai/reference#rest-api-versioning
Ref: https://github.com/openai/openai-python/blob/17ac6779958b2b74999c634c4ea4c7b74906027a/src/openai/lib/azure.py#L170

**Comment by miararoy** [comment]
Thanks! 
Just letting you know we have internally started to work on this as well, we will make sure to merge the work so this could be shipped asap

File: src/canopy/llm/openai.py
Comment (igiloh-pinecone) [inline src/canopy/llm/openai.py:240]:
```suggestion
        super().__init__(model_name=None)
```
There's not point sending the `model_name` to `super().__init__` and save it under `self.model_name`.  
In the `chat.completion.create()` call, the `model=` argument is interpreted as the `deployment` anyway (and ignored if `deployment` was explicitly defined).  
In the context of `AzureOpenAILLM`, the init arg `model_name` should be synonym with `model deployment`. Of course the docstring would need to explain that carefully.

File: src/canopy/llm/openai.py
Comment (igiloh-pinecone) [inline src/canopy/llm/openai.py:244]:
```suggestion
            azure_endpoint=base_url,
            azure_deployment=model_name
```
See comment above about using `self.model_name`

File: src/canopy/llm/openai.py
Comment (igiloh-pinecone) [inline src/canopy/llm/openai.py:312]: A. Are we sure that if you deploy your model with the latest GPT version (e.g. `gpt-3.5-turbo-1106`), the `tools=` argument is still not supported?  
If it does - we should probably remove the duplication and use `OpenAILLM`'s implementation directly.

Comment (igiloh-pinecone) [reply src/canopy/llm/openai.py:312]: B. I noticed that the default `gpt-3.5-turbo` version on Azure OpenAI services is `0301`, which doesn't support function calling. It means that for many users - this code is going to error out, and OpenAI's error message is not very intuitive.  

We should probably add a `try: except openai.NotFoundError:` clause around this call, and re-raise with a clearer error message that tells the user that the model version defined in their deployment doesn't support function calling.

File: src/canopy_server/app.py
Comment (igiloh-pinecone) [inline src/canopy_server/app.py:21]: @MichaelAnckaert for this PR you can leave these changes as-is, but in the future please avoid running linters like `black` on files (or even lines) that weren't  changed in your PRs. This results in redundant conflicts with other PRs.

File: src/canopy/llm/openai.py
Comment (igiloh-pinecone) [inline src/canopy/llm/openai.py:323]: Note that the `.available_models` property is broken, as it seems that Azure OpenAI API doesn't allow listing all of your deployments (==models).  
We should wrap that with a clear error message, and point to where the list of available deployments can be found in the Azure console

File: src/canopy/knowledge_base/record_encoder/openai.py
Comment (igiloh-pinecone) [inline src/canopy/knowledge_base/record_encoder/openai.py:55]: Last but not least - please note that this class requires its own unit test before it can be merged.

File: src/canopy/llm/openai.py
Comment (igiloh-pinecone) [inline src/canopy/llm/openai.py:341]: Note that the `AzureOpenAILLM` class requires its own system test (since we don't have stand alone unit test for `LLM` classes yet).  
The test can assume that an Azure OpenAI endpoint was created, and that both the endpoint and the API key are provided by env vars.  
Once the test file exists (and passes locally on your branch) - we will add `AZURE_OPENAI_API_KEY` and `AZURE_OPENAI_ENDPOINT` secretes to the repo's CI configuration.

File: src/canopy/llm/openai.py
Comment (MichaelAnckaert) [reply src/canopy/llm/openai.py:223]: Makes me wonder if I should change the implementation of `AzureOpenAILLM` to have different init arguments from the default `OpenAILLM` class? I added the `api_version` argument, does it make sense to rename `model_name` to `deployment_name`?

Comment (igiloh-pinecone) [reply src/canopy/llm/openai.py:223]: @MichaelAnckaert `model_name` is part of the signature of all `BaseLLM` inheritors, so that is the only parameter that has to still exist in order to comply with `BaseLLM` interface.  
You can add additional init arguments or even change\remove ones that come from `OpenAILLM`, as long as you maintain compatibility with  `BaseLLM`.

**Comment by igiloh-pinecone** [comment]
@MichaelAnckaert there hasn't been any activity in this PR since the review two weeks ago.  
Please advise if you are planning to finalize this PR so it can be merged.

**Comment by igiloh-pinecone** [comment]
@MichaelAnckaert Thank you very much for your contribution!!

Since this PR has been inactive, we have finalized it with all missing tests and prepared it for merging.  
All of your original commits are included of course, and will be merged to Canopy.

**Review by acatav — DISMISSED** [review]
LGTM overall, look at the minor suggestions. I mainly think we should decide if api version goes to env vars or config, and can consider enable model name in config at some sense

**Review by igiloh-pinecone — APPROVED** [review]
LGTM!

**Pull request merged**
